### DISCUSSION TITLE
New Documentation Website

### DISCUSSION THREAD

**Discussion by krainboltgreene** [vcr/vcr#856] [Ideas]
#712 #304 #17 #694 #525

AKA "Relish is dead, what now?"

Apparently they have a new site called https://cucumber.io/jam.

**Comment by andrewmcodes** [comment]
@krainboltgreene you could use a gem to convert the cucumber files to markdown and then use something like [gitbook](https://www.gitbook.com/) or roll a site with Gatsby or Jekyll. I’d be happy to help with this.

**Comment by krainboltgreene** [comment]
I've used gitbook extensively in the past, I like them well enough. Sure, if you're willing to devote time to it then make the PR, you'll get my primary attention.

**Comment by olleolleolle** [comment]
Going to try a few tools, see what they're made of.

http://www.picklesdoc.com/ for example. (I learned: It's a Windows tool, and I don't run that.)

**Comment by andrewmcodes** [comment]
I used [this tool](https://github.com/raviqqe/gherkin2markdown) to see the type of markup it would produce and [here](https://app.gitbook.com/@andrewmcodes/s/vcr/v/v5.0.0/) is an example of 5.0 docs on gitbook. I noticed it strips the language tags from code blocks, omits sections that use `@exclude-foo`, and a few indentation issues. @krainboltgreene if you think it’s fine, I can port over all the docs and discuss how to automate or explore another tool. I really don’t want to write my own conversion gem unless I have to but I am personally not a fan of the output. I have looked at other tools and haven’t found a good alternative and wasn’t successful at my first attempt to create a conversion gem 🙄

**Comment by olleolleolle** [comment]
Blog post we could link to: https://fabioperrella.github.io/10_tips_to_help_using_the_VCR_gem_in_your_ruby_test_suite.html

**Comment by dgarwood** [comment]
@andrewmcodes @krainboltgreene appears their Jam got canned as well: https://smartbear.com/blog/test-and-monitor/smartbear-acquires-cucumber/

looks like this hasn't been updated in a while.  I wasn't able to view the example on gitbook.  Is there anything I can do to help?

**Comment by olleolleolle** [comment]
_Any_ research about how to create Relish-like output are interesting to this Issue.

@andrewmcodes This is what the gitbook link looked like on logging in:

<img width="492" alt="bild" src="https://user-images.githubusercontent.com/211/80573906-26df7300-8a01-11ea-8240-be658ab3ce35.png">


✨

**Comment by nicolasiensen** [comment]
I decided to give it a try using [gherkin2markdown](https://github.com/raviqqe/gherkin2markdown) and [GitBook](https://www.gitbook.com/), just like @andrewmcodes suggested.

I tried to keep the same structure as we have in [Relishapp](https://relishapp.com/vcr/vcr/docs), but I didn't create all the pages as my goal here is to have a proof of concept.

I imported all the pages manually, but I'm pretty sure we could automate this process if we decide to go with this solution.

The pages from Getting Started to Contributing are all markdown files contained in the source code of the project. The pages under the Test Frameworks section were transformed by gherkin2markdown and then imported to GitBook.

Let me know what you think.

https://nicolasiensen.gitbook.io/vcr-documentation/

**Comment by olleolleolle** [comment]
It looks like a good start.

Thanks for taking this up, @nicolasiensen! 

An enjoyable presentation.

(After this line, nonproductive bike-shedding! 😄 )

<details>

---

Some (future) polishing for readability would make things shinier (syntax highlighting).
Some typography tweaks like in the Relish example below:

<img width="738" alt="bild" src="https://user-images.githubusercontent.com/211/83339975-6e447200-a2d3-11ea-9149-88227170a7b3.png">

What I like: the structure is strongly presented with some justification/whitespace.

</details>

**Comment by nicolasiensen** [comment]
I gave it another try using a much less manual approach:

We add a `SUMMARY.md` file in the root folder of the repository containing the structure of the VCR documentation, as [documented by GitBook](https://docs.gitbook.com/integrations/github/content-configuration#summary).

You can check how the codebase would look like on [this branch](https://github.com/vcr/vcr/compare/master...nicolasiensen:configure-gitbook).

Following this approach, we would need to include the markdown version of the feature files as part of the codebase 🙁.

@olleolleolle I didn't tackle the feedback you provided above just yet.

Not much has changed from the previous trial, but here is how the documentation looks like https://nicolasiensen.gitbook.io/vcr

**Comment by nicolasiensen** [comment]
I've worked a bit more on it today, and I managed to add GitBook variants to account for different versions of the library:

![vcr-doc](https://user-images.githubusercontent.com/208312/83411556-c6d15780-a418-11ea-999c-056e9adddf68.gif)

We can setup GitBook to create one variant of the documentation per branch. Unfortunately, it doesn't work with git tags.

We can specify the branches we want GitBook to create documentation variants. I configured it to catch all branches matching the pattern `v-*`.

I created two branches on my fork derived from their respective tags:
- https://github.com/nicolasiensen/vcr/tree/v-6.0.0
- https://github.com/nicolasiensen/vcr/tree/v-5.1.0

I pushed to each branch the files necessary to integrate with GitBook and voilà.

I have to say that it would be quite some work to set up the documentation for each version of the library, but it is possible to do it.

I have also completed the documentation, so it has all the sections we currently have in Relishapp.

You can check the documentation in the following link:

https://nicolasiensen.gitbook.io/vcr/

**Comment by nicolasiensen** [comment]
I was reading about what the folks from RSpec are planning to do regarding the discontinuation of Relish, and it seems that they are planning to migrate their documentation to Cucumber Pro: https://github.com/rspec/rspec-dev/issues/199.

GitBook is cool, but the Gherkin documentation would never look as gorgeous as if we host it in a platform built for that purpose. Also, the tooling around the project to transform Gherkin files into Markdown would add an extra load to the project ecosystem that I'm not a big fan of.

All in all, I would wait to see what Cucumber Pro is going to turn into. It looks like Relish [will stay around](https://github.com/rspec/rspec-rails/issues/2065#issuecomment-456157048) as long as Cucumber Pro is not ready for RSpec.

What do you think?

**Comment by olleolleolle** [comment]
Looks like Jam, aka Cucumber Pro, is a ways off in the future.

**Comment by andrewmcodes** [comment]
I totally let this slip off my radar after changing teams. Unfortunately I just don't have the bandwidth to pick it back up. Glad to see some folks were able to step up 🙏

**Comment by nicolasiensen** [comment]
I have set the GitBook project aside for a bit to explore how we could leverage our existing YARD documentation to include the Cucumber features.

I found a project called [`yard-cucumber`](https://github.com/burtlo/yard-cucumber), which is not very active but does a fair job transforming each feature file into a YARD documentation page and indexing it to be searchable.

This way, we centralize the documentation of the library into one single place while we stay away from the extra tooling that GitBook requires.

You can have a sneak peek of the result below:

![vcr](https://user-images.githubusercontent.com/208312/91759510-06533f80-ebd2-11ea-94d2-1d888e4a79e0.gif)

There is certainly room for improvement, but I wanted to hear what you think about this approach before investing more time on it.

**Comment by olleolleolle** [comment]
Possibility: YARD-generated HTML could be served as GitHub Pages content. `yard-cucumber` looks very strong within the constraints given.

**Comment by olleolleolle** [comment]
To get close to what we had on Relish, perhaps we can use the cucumber HTML presenter:

- Run individual scenarios as standalone runs with the HTML formatter
- Publish multiple pages under the v6 namespace
- Individually link

**Comment by mattwynne** [comment]
Folks, I just wanted to add a bit of perspective to this from the "inside" of Cucumber/SmartBear.

We've very much committed to supporting open source BDD teams publish their living documentation, and I'm sorry we haven't yet managed to achieve feature parity in the regard with Relish on any of the other attempts we've made to solve this problem (Cucumber Pro, Jam, Cucumber Reports, and CucumberStudio).

We have stopped allowing sign-ups on https://relishapp.com as we don't see a _commercial_ future for the project/codebase, and don't want to burn hours on support and maintenance, but we'll keep it up and running to host RSpec, VCR and other public projects' docs as long as there isn't a better alternative. If an open source project like this one needs users to have access, we can still do that from the Rails console, so just let me know. Consider yourselves as having VIP level support status 😁 

I'll cc @cbliard and @tooky here who are in charge of the priorities for https://reports.cucumber.io and https://studio.cucumber.io and can bear this need in mind as they're making plans.

**Reply by olleolleolle** [reply]
A couple of things happened:

- @mattwynne fixed me up with an account and attached me to the vcr Relishapp project
- I reran `NEW_RELISH_RELEASE=true bundle exec rake relish`
- 6.0.0 is now live, please take a look and see if https://relishapp.com/vcr/vcr/v/6-0-0/docs looks alright
